app/scripts/controllers/permissions/index.js

```diff
@@ -510,6 +507,119 @@ export class PermissionsController {
     })
   }
 
+  /**
+   * Removes all known domains and their related permissions.
+   */
+  clearPermissions () {
+    this.permissions.clearDomains()
+    this.notifyAllDomains({
+      method: NOTIFICATION_NAMES.accountsChanged,
+      result: [],
+    })
+  }
+
+  /**
+   * Stores domain metadata for the given origin (domain).
+   * Deletes metadata for domains without permissions in a FIFO manner.
+   * Metadata is only deleted for domains with permissions on boot in order to
+   * prevent a degraded user experience, as metadata currently can't be
+   * requested on-demand.
+   *
+   * @param {string} origin - The origin whose domain metadata to store.
+   * @param {Object} metadata - The domain's metadata that will be stored.
+   */
+  addDomainMetadata (origin, metadata) {
+
+    const oldMetadataState = this.store.getState()[METADATA_STORE_KEY]
+    const newMetadataState = { ...oldMetadataState }
+
+    if (Object.keys(newMetadataState).length >= METADATA_STORE_MAX_SIZE) {
+      this._popDomainMetadata(newMetadataState)
+    }
+
+    newMetadataState[origin] = {
+      ...oldMetadataState[origin],
+      ...metadata,
+      lastUpdated: Date.now(),
+    }
+    this._newMetadataOrigins.add(origin)
+
+    this._setDomainMetadata(newMetadataState)
+  }
+
+  /**
+   * Deletes metadata for domains without permissions when the metadata store
+   * has grown too large. Highly dependent on logic of addDomainMetadata.
```

Comment:
>Highly dependent on logic of addDomainMetadata

Could this be inlined in `addDomainMetadata` instead then? The behavior of this function is.... pretty strange. I don't think it makes sense as an independent function.

Reply:
Done in faaa60b